src/Storages/VirtualColumnUtils.cpp

```diff
@@ -144,26 +148,74 @@ NameSet getVirtualNamesForFileLikeStorage()
+std::pair<ColumnPtr, ColumnPtr> parseHivePartitioningKeysAndValues(const String & path)
 {
-    std::string pattern = "([^/]+)=([^/]+)/";
-    re2::StringPiece input_piece(path);
-
-    std::unordered_map<std::string, std::string> key_values;
-    std::string key;
-    std::string value;
-    std::unordered_map<std::string, std::string> used_keys;
-    while (RE2::FindAndConsume(&input_piece, pattern, &key, &value))
+    static auto extractor = KeyValuePairExtractorBuilder()
+    .withItemDelimiters({'/'})
+    .withKeyValueDelimiter('=')
+    .build();
+
+    auto keys = ColumnString::create();
+    auto values = ColumnString::create();
+
+    // cutting the filename to prevent malformed filenames that contain key-value-pairs from being extracted
+    // not sure if we actually need to do that, but just in case. Plus, the previous regex impl took care of it
+    const auto last_slash_pos = path.find_last_of('/');
+
+    if (last_slash_pos == std::string::npos)
     {
-        auto it = used_keys.find(key);
-        if (it != used_keys.end() && it->second != value)
-            throw Exception(ErrorCodes::INCORRECT_DATA, "Path '{}' to file with enabled hive-style partitioning contains duplicated partition key {} with different values, only unique keys are allowed", path, key);
-        used_keys.insert({key, value});
+        // nothing to extract, there is no path, just a filename
+        return std::make_pair(std::move(keys), std::move(values));
+    }
+
+    std::string_view path_without_filename(path.data(), last_slash_pos);
+
+    extractor->extract(path_without_filename, keys, values);
+
+    keys->validate();
+    values->validate();
+
+    std::unordered_set<StringRef> check_for_duplicates_set;
 
-        auto col_name = key;
-        key_values[col_name] = value;
+    for (std::size_t i = 0; i < keys->size(); i++)
+    {
+        const auto key = keys->getDataAt(i);
+        const auto [_, inserted] = check_for_duplicates_set.insert(keys->getDataAt(i));
+
+        if (!inserted)
+        {
+            throw Exception(
+                ErrorCodes::INCORRECT_DATA,
+                "Path '{}' to file with enabled hive-style partitioning contains duplicated partition key {}, only unique keys are allowed",
+                path,
+                key.toString());
+        }
     }
-    return key_values;
+
+    return std::make_pair(std::move(keys), std::move(values));
+}
+
+std::optional<std::pair<String, String>> getKeyAndValuePairFromHiveKeysAndValues(const std::optional<std::pair<ColumnPtr, ColumnPtr>> & hive_keys_and_values_opt, const String & key)
+{
+    if (!hive_keys_and_values_opt)
+    {
+        return std::nullopt;
+    }
+
+    const auto keys_column = hive_keys_and_values_opt.value().first;
+    const auto values_column = hive_keys_and_values_opt.value().second;
+
+    for (std::size_t i = 0; i < keys_column->size(); i++)
```

Comment:
Ok, CHKeyValuePairExtractor is faster than regexp.
But here key is finding in cycle (O(N)) instead of search in unordered map (O(1)) before this PR.
Does it really work faster?

Reply:
This column, in the majority of cases, will contain only 1~3 rows. It's very unlikely to have a table partitioned by more columns.

Being that said, technically, the column lookup is indeed  O(N), but for this constrained input size (1–3 items), it behaves as O(1) in practice.

Please have a look at the benchmarks I shared here: https://github.com/ClickHouse/ClickHouse/pull/79067